src/xml_parser_main_metadata_together.c

```diff
@@ -432,15 +432,27 @@ int cr_xml_parse_main_metadata_together(const char *primary_path,
         }
     }
 
-    cr_xml_parser_data_free(primary_pd);
-    cr_xml_parser_data_free(filelists_pd);
-    cr_xml_parser_data_free(other_pd);
+    // When interrupted at the right time primary_pd->pkg can either be:
+    //  - referenced in both primary_pd->pkg and cbdata.in_progress_pkgs_list or
+    //  - referenced only in primary_pd->pkg (we have started parsing the pkg but didn't finish it yet)
+    // in order to avoid a crash remove it from the list if present
+    // primary_pd->pkg is a special case because of newpkgcb_primary
+    if (cbdata.in_progress_pkgs_list) {
+        cbdata.in_progress_pkgs_list = g_slist_remove(cbdata.in_progress_pkgs_list, primary_pd->pkg);
+    }
+    if (primary_pd) {
+        cr_package_free(primary_pd->pkg);
+    }
```

Comment:
IIRC (it has been a month since I last looked at all of this) that it gets more complicated when the Python bindings are involved, because there can be other indirect references via https://github.com/rpm-software-management/createrepo_c/blob/master/src/python/xml_parser-py.c#L37 and via the PyObjects themselves, and having the C code free the package can lead to problems when the package PyObjects get DECREF'd and try to free the package payload (again).